bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java

```diff
@@ -234,48 +245,116 @@ private TomlNode getTomlNode(TomlNode tomlValue, String variableName, Type type)
+        String moduleName = module.getName();
+        TomlNode errorNode = toml.rootNode();
+        Optional<TomlValueNode> valueNode = toml.get(moduleName);
+        List<Toml> tomlTables = toml.getTables(moduleName);
+        if (valueNode.isEmpty()) {
+            valueNode =  toml.get(getModuleKey(module));
+        }
+        if (tomlTables.isEmpty()) {
+            tomlTables =  toml.getTables(getModuleKey(module));
+        }
+        if (valueNode.isPresent()) {
+            errorNode = valueNode.get();
+        } else if (!tomlTables.isEmpty()) {
+            errorNode = tomlTables.get(0).rootNode();
+        } else {
+            Optional<Toml> tomlValueNode = toml.getTable(moduleName.replaceFirst(rootModule.getName() + ".", ""));
+            errorNode = tomlValueNode.isPresent() ? tomlValueNode.get().rootNode() : errorNode;
         }
-        return (TomlTableNode) retrievedNode;
+        invalidRequiredModuleSet.add(module.toString());
+        throw new TomlConfigException(String.format(INVALID_MODULE_STRUCTURE, moduleName, moduleName), errorNode);
     }
 
-    private TomlTableNode extractModuleNode(TomlTableNode orgNode, String moduleName, String fullModuleName,
-                                                   boolean hasRequired) {
-        if (orgNode == null) {
-            return null;
+    private TomlTableNode getRootModuleNode(Toml baseToml) {
+        String moduleName = rootModule.getName();
+        String moduleKey = getModuleKey(rootModule);
+        Optional<Toml> table = baseToml.getTable(moduleKey);
+        if (table.isEmpty()) {
+            if (hasModuleAmbiguity) {
+                throw new TomlConfigException(String.format(CONTAINS_MODULE_AMBIGUITY, moduleName, moduleKey),
+                        baseToml.rootNode());
+            }
+            table = baseToml.getTable(moduleName);
+            if (table.isEmpty() || subModuleSet.containsAll(table.get().rootNode().entries().keySet())) {
+                return baseToml.rootNode();
+            }
         }
-        TomlTableNode moduleNode = orgNode;
-        int subModuleIndex = moduleName.indexOf(SUBMODULE_DELIMITER);
-        if (subModuleIndex == -1) {
-            moduleNode = validateAndGetModuleStructure(orgNode, moduleName, fullModuleName);
-            if (moduleNode == null && hasRequired && !invalidRequiredModuleSet.contains(fullModuleName)) {
-                invalidRequiredModuleSet.add(fullModuleName);
-                throw new TomlConfigException(String.format(INVALID_MODULE_STRUCTURE, fullModuleName,
-                        fullModuleName), orgNode);
+        return table.map(Toml::rootNode).orElse(null);
+    }
+
+    private void analyseModuleAmbiguity() {
+        for (Module entry : moduleSet) {
+            String rootModuleName = rootModule.getName();
+            if (rootModuleName.startsWith(entry.getOrg())) {
+                hasModuleAmbiguity = true;
```

Comment:
Will this make all the modules to be configured with org name if one module name has clashed with org name. Should we keep this per module.

Reply:
This check considers the clash between org-name and root module-name. We mandate the org-name in this case. 
Therefore, if there's a clash then it mandates all the modules (as the non-default modules also starts with root module-name & org-name is always mandatory for imported packages)

So, I think we can keep this as a boolean.